### PR TITLE
Entity company

### PR DESCRIPTION
**Pull request by chu4**
Company model with controller and assosiations to user
+basic views (using simple_form for new and edit views)
+tests for the model and controller respectively(with factories)

companies_edit
companies_index
companies_new
companies_show

### CODE REVIEW

**Comment by igorkasyanchuk** [comment]
move it all to "/account/" namespace as we discussed.

File: app/assets/javascripts/companies.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/companies.js]: remove this file

File: app/controllers/companies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+class CompaniesController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @companies = Company.all
```
Comment (igorkasyanchuk) [inline app/controllers/companies_controller.rb]: so as any user I can see all companies?

File: app/controllers/companies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  private
+
+  def company_params
+    params.require(:company).permit(:name, :domain)
+  end
+
+  def current_company
+    Company.find(params[:id])
```
Comment (igorkasyanchuk) [inline app/controllers/companies_controller.rb]: as any user, I can open any company? Just put a different ID in URL

PS: use current_user.companies....

File: config/routes.rb [hunk 4]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   devise_for :users
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  resources :companies
```
Comment (igorkasyanchuk) [inline config/routes.rb]: move to account namespace

File: spec/controllers/companies_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,151 @@
+require 'rails_helper'
+
+RSpec.describe CompaniesController, type: :controller do
+  let(:user) { create :user }
```
Comment (igorkasyanchuk) [inline spec/controllers/companies_controller_spec.rb]: always use "let!" instead of "let"

File: spec/controllers/companies_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    it "renders the :index view" do
+      get :index
+      expect(response).to have_http_status(200)
+      expect(response).to render_template(:index)
+    end
+  end
+
+  describe "GET #show" do
+    let!(:company) { create :company }
```
Comment (igorkasyanchuk) [inline spec/controllers/companies_controller_spec.rb]: create :company change to create(:company)
and in other places too

File: app/assets/javascripts/companies.js (hunk 1 above)
Comment (chu4) [reply app/assets/javascripts/companies.js]: removed as well as companies.scss

File: config/routes.rb (hunk 4 above)
Comment (chu4) [reply config/routes.rb]: Added account namespace and refactored routes, controller, views and tests with new path helpers

File: spec/controllers/companies_controller_spec.rb (hunk 5 above)
Comment (chu4) [reply spec/controllers/companies_controller_spec.rb]: changed it, but in this case it was 'let' cause user was only used in 5th line for signing_in before each test

**Comment by chu4** [comment]
Improved code, respectively to the suggested changes

File: app/controllers/account/companies_controller.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
+module Account
+  class CompaniesController < ApplicationController
```
Comment (igorkasyanchuk) [inline app/controllers/account/companies_controller.rb]: change to Account::CompaniesController

Comment (igorkasyanchuk) [reply app/controllers/account/companies_controller.rb]: and remove module Account

File: app/controllers/account/companies_controller.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    before_action :authenticate_user!
+
+    def index
+      @companies = Company.all
```
Comment (igorkasyanchuk) [inline app/controllers/account/companies_controller.rb]: current_user.companies.all

File: app/controllers/account/companies_controller.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    private
+
+    def company_params
+      params.require(:company).permit(:name, :domain)
+    end
+
+    def current_company
+      Company.find(params[:id])
```
Comment (igorkasyanchuk) [inline app/controllers/account/companies_controller.rb]: current_user.companies.find

File: app/controllers/account/companies_controller.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    def edit
+      @company = current_company
+    end
+
+    def update
+      @company = current_company
+      if @company.update_attributes(company_params)
+        redirect_to account_company_path
```
Comment (romaslmd) [inline app/controllers/account/companies_controller.rb]: is this correct path? i believe for resources it should be either `account_companies_path` or `account_company_path(@company)`

Comment (romaslmd) [reply app/controllers/account/companies_controller.rb]: some flash message?

File: app/helpers/companies_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module CompaniesHelper
```
Comment (romaslmd) [inline app/helpers/companies_helper.rb]: incorrect module location
is it really needed?

File: spec/controllers/account/companies_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+RSpec.describe Account::CompaniesController, type: :controller do
+  let!(:user) { create(:user) }
+  before(:each) { sign_in user }
+
+  describe "GET #index" do
+    let!(:company) { create(:company) }
+
+    it "populates an array of companies" do
+      get :index
+      expect(assigns(:companies)).to eq([company])
```
Comment (romaslmd) [inline spec/controllers/account/companies_controller_spec.rb]: never do like this
`ActiveRecord:Association_Collection` is not an array
prefer to use `RSpec::Matchers#contain_exactly`

File: app/helpers/companies_helper.rb (hunk 11 above)
Comment (chu4) [reply app/helpers/companies_helper.rb]: true, as of the moment there is no need in that helper, so i removed it

File: app/controllers/account/companies_controller.rb (hunk 10 above)
Comment (chu4) [reply app/controllers/account/companies_controller.rb]: 'account_company_path' works for updating but doesn't for creating, so there i'll use 'account_company_path(@company)' as suggested.
will add flashes, thanks

**Comment by chu4** [comment]
Refactored with suggested changes. Added Flash messages. Updated with current master branch.

flash message example
image

File: app/views/layouts/application.html.slim [hunk 13]
```diff
@@ -11,4 +11,6 @@ html
       = notice
     p.alert
       = alert
```
Comment (igorkasyanchuk) [inline app/views/layouts/application.html.slim]: what is "=alert"? above. And now you have added flash.each? remove line #13 probably

File: spec/controllers/account/companies_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,143 @@
+require 'rails_helper'
+
+RSpec.describe Account::CompaniesController, type: :controller do
+  let!(:company) { create(:company) }
+  before(:each) { sign_in User.last }
```
Comment (igorkasyanchuk) [inline spec/controllers/account/companies_controller_spec.rb]: why User.last? This is wrong. Use owner of a company.

Comment (igorkasyanchuk) [reply spec/controllers/account/companies_controller_spec.rb]: and remember ... in many cases when you have code like ".first", ".last" there is a bug

File: spec/models/company_spec.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe Company, type: :model do
+  describe 'validations with factories' do
+    let(:val_company) { build :company }
```
Comment (igorkasyanchuk) [inline spec/models/company_spec.rb]: use "let!" always
and "build(:company)" with brackets

File: app/views/layouts/application.html.slim (hunk 13 above)
Comment (chu4) [reply app/views/layouts/application.html.slim]: that line was generated by default, removed

File: spec/controllers/account/companies_controller_spec.rb (hunk 14 above)
Comment (chu4) [reply spec/controllers/account/companies_controller_spec.rb]: My bad, company.user should be a better choice

**Pull request merged**